### ISSUE TITLE
Autofac and Pipeline Behavior

### ISSUE THREAD

**Issue by martosource** [jbogard/MediatR#128]
I am trying to get new pipeline working with autofac but it is running the same pipeline many times with different types.

Adding the following class the MediatR example

```
public class BehaviorOne<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse> 
    {

        private readonly TextWriter _writer;

        public BehaviorOne(TextWriter writer)
        {
            _writer = writer;
        }

        public async Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TResponse> next)
        {
            //System.Diagnostics.Trace.WriteLine($"BehaviorOne - Handle {typeof(TRequest).Name}");

            _writer.WriteLine("BehaviorOne - Handle " + typeof(TRequest).Name);

            var response = await next();

            return response;
        }
    }
```

Added this to the BuildMediator for the autofac example program.cs

    builder.RegisterGeneric(typeof(BehaviorOne<,>)).As(typeof(IPipelineBehavior<,>));

Produced the following output:

```
Sample mediator implementation using send, publish and post-request handlers in sync and async version.
---------------
Sending Ping...
BehaviorOne - Handle Ping
BehaviorOne - Handle Object
BehaviorOne - Handle IRequest`1
Received: Ping Pong
Sending Ping async...
BehaviorOne - Handle PingAsync
BehaviorOne - Handle Object
BehaviorOne - Handle IRequest`1
Received async: Ping Pong
Publishing Pinged...
Got pinged also.
Got pinged.
Got notified.
Got notified also async.
Publishing Pinged async...
Got notified.
Got pinged also async.
Got pinged async.
Got notified also async.
```

For the structure map example you get:

```
Sample mediator implementation using send, publish and post-request handlers in sync and async version.
---------------
Sending Ping...
BehaviorOne - Handle Ping
Received: Ping Pong
Sending Ping async...
BehaviorOne - Handle PingAsync
Received async: Ping Pong
Publishing Pinged...
Got notified.
Got pinged.
Got pinged also.
Got notified also async.
Publishing Pinged async...
Got notified.
Got notified also async.
Got pinged async.
Got pinged also async.
```

**Comment by JamesFenton** [comment]
I am having the same issue. The pipeline's Handle method gets called 3 times after sending Ping as @martosource describes.

Removing the line
```
builder.RegisterSource(new ContravariantRegistrationSource());
```
seems to fix the problem

**Comment by martosource** [comment]
@JamesFenton all fixed, thanks.

**Comment by JamesFenton** [comment]
It seems Autofac does not support registering open generics when doing assembly scanning: https://github.com/autofac/Autofac/issues/215

What this means is that the built in pipeline behaviors `RequestPreProcessorBehavior` and `RequestPostProcessorBehavior` do not get registered from the following line from the Autofac sample:
```
builder.RegisterAssemblyTypes(typeof (IMediator).GetTypeInfo().Assembly).AsImplementedInterfaces();
```
This also means that any open generic `IRequestPreProcessor` or `IRequestPostProcessor` implementations will not get registered, for example:
```
public class GenericPreProcessor<TRequest> : IRequestPreProcessor<TRequest>
    {
        public Task Process(TRequest request)
        {
            return Task.Delay(0);
        }
    }
```
The built-in behaviours can be registered using
```
builder.RegisterGeneric(typeof(RequestPreProcessorBehavior<,>)).AsImplementedInterfaces();
builder.RegisterGeneric(typeof(RequestPostProcessorBehavior<,>)).AsImplementedInterfaces();
```
But registering of all open generic `IRequestPreProcessor` or `IRequestPostProcessor` implementations using assembly scanning requires some more specialised code, possibly similar to https://github.com/autofac/Autofac/issues/215#issuecomment-248538820

**Comment by adamhathcock** [comment]
I switched to using Autofac instead of using the ASP.NET Core default container because Autofac supports better generic constraints.

I wanted to use interfaces as markers for certain pipeline behaviors like this and Autofac does it:
```csharp
public class DistributedCachingHandler<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
        where TRequest : class, IUseDistributedCache
        where TResponse : class
```

Anyway, I believe I would see this issue but I don't because I'm still using the `IServiceCollection` way of registering types instead of the Autofac builder.  I believe it's confirmed above from @JamesFenton 

Maybe using `IServiceCollection` registration and assembly scanning for Autofac is an option.

**Comment by JamesFenton** [comment]
What I would like to do is write some tests for each of the different containers. Ideally I'd like to keep the registration code in each of the samples so that people can easily find it. I'd also like to avoid "infecting" the current test project with all the different containers, so perhaps another test project that references all the same projects? What do you think @jbogard?

**Comment by jbogard** [comment]
Well we do have the example projects, I wouldn't mind extending those perhaps to highlight what capabilities each container has through tests.

**Comment by jbogard** [comment]
Yeah that sounds good to me!

**Comment by barryschulz** [comment]
@JamesFenton If you do not register ContravariantRegistrationSource, it seems that you will lose out on the polymorphic dispatch. For example, if you had an implementation of` IRequestPreProcessor<IPing>` and Ping implemented IPing, this processor would not be found with the current `RequestPreProcessorBehavior<TRequest,TResponse>`.

Are you seeing this same limitation?

**Comment by worldspawn** [comment]
So was there a solution without caveats?

**Comment by jbogard** [comment]
You can look at the way StructureMap does this and borrow its implementation?

**Comment by worldspawn** [comment]
I should have checked the autofac issue, just reposting this here for posterity, a suggested workaround for autofac is:

```
builder.RegisterAssemblyTypes(assemblies)
    .Where(t => t.GetTypeInfo()
        .ImplementedInterfaces.Any(
            i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof (IRepository<>)))
    .AsImplementedInterfaces()
    .InstancePerLifetimeScope();
```

Swapping IRepository for IPipelineBehavior i guess.  Havent tried it yet.

**Comment by JamesFenton** [comment]
@barryschulz you are correct, if you remove the `ContravariantRegistrationSource`, you cannot use contravariance, e.g. IRequestPreProcessor<IPing>, but if you keep it, you cannot use generic pipeline behaviours. We haven't needed to use contravariance, but we have used generic pipeline behaviours to process every request. I suppose that achieves the same thing as contravariance, provided your container supports generic constraints like @adamhathcock suggested.

The setup of Autofac does seem to require a lot more ceremony than StructureMap, unfortunately.

**Comment by seroche** [comment]
Hey guys, 

Very old topic but I had the same problem this morning and I solved by implementing my own ScopedContravariantRegistrationSource (as advised by jdaigle). It's a very simple implementation but it works quite well with my very easy scenario:

```
    public class ScopedContravariantRegistrationSource : IRegistrationSource
    {
        private readonly IRegistrationSource _source = new ContravariantRegistrationSource();
        private readonly List<Type> _types = new List<Type>();

        public ScopedContravariantRegistrationSource(params Type[] types)
        {
            if (types == null)
                throw new ArgumentNullException(nameof(types));
            if (!types.All(x => x.IsGenericTypeDefinition))
                throw new ArgumentException("Supplied types should be generic type definition");
            _types.AddRange(types);
        }

        public IEnumerable<IComponentRegistration> RegistrationsFor(
            Service service,
            Func<Service, IEnumerable<IComponentRegistration>> registrationAccessor)
        {
            var components = _source.RegistrationsFor(service, registrationAccessor);
            foreach (var c in components)
            {
                var defs = c.Target.Services
                    .OfType<TypedService>()
                    .Select(x => x.ServiceType.GetGenericTypeDefinition());

                if (defs.Any(_types.Contains))
                    yield return c;
            }
        }

        public bool IsAdapterForIndividualComponents => _source.IsAdapterForIndividualComponents;
    }
```

**Comment by maldworth** [comment]
@seroche Can you give an example of usage as well?

**Comment by bilauca** [comment]
I have the same behavior on behaviors as @martosource or @JamesFenton described on dotnet core default DI. Any hints on how to fix it without my own error prone ScopedContravariantRegistrationSource?

**Comment by billkiddo** [comment]
I'm using the recommended ASP.NET Core DI registration, see below
`
            services.AddScoped<ServiceFactory>(p => p.GetService);
            services.AddScoped(typeof(IPipelineBehavior<,>), typeof(RequestPreProcessorBehavior<,>));
            services.Scan(scan => scan
                .FromAssembliesOf(typeof(IMediator), typeof(FeasibilityApplicationService))
                .AddClasses()
                .AsImplementedInterfaces());
`
Removing the Scrutor part, seems to do the trick for now.

**Comment by jesulink2514** [comment]
@maldworth I guess you should change your ContravariantRegistrationSource to this 

`builder.RegisterSource(new ScopedContravariantRegistrationSource ());`

**Comment by cyrildurand** [comment]
I had a similar issue with MediatR and `INotificationHandler` I finally doing my own registration source that extend the original ContravariantRegistrationSource 

    /// <summary>
    /// Returns contravariant registration source without duplicator target. 
    /// 
    /// <see cref="ContravariantRegistrationSource" /> returns all contravariant implementation of a type. 
    /// For example when we resolve IEnumerable&lt;INotificationHandler&lt;SpecificCommand&gt;&gt; it will returns a collection with GenericHandler&lt;SpecificCommand&gt;, GenericHandler&lt;BaseCommand&gt;, SpecificCommandHandler 
    /// this registration source will first look up for the native registrations and then group registration based on activator limit type. 
    /// </summary>
    /// <remarks>See https://stackoverflow.com/questions/46464944/autofac-contravariance-and-resolving-open-generic-types </remarks>
    public class ExplicitContravariantRegistrationSource : IRegistrationSource
    {
        private readonly IRegistrationSource _source = new ContravariantRegistrationSource();
        private readonly Type _type;

        public ExplicitContravariantRegistrationSource(Type type)
        {
            if (type == null)
            {
                throw new ArgumentNullException(nameof(type));
            }
            if (!type.IsGenericTypeDefinition)
            {
                throw new ArgumentException("Type should be a generic type definition", nameof(type));
            }
            this._type = type;
        }

        public IEnumerable<IComponentRegistration> RegistrationsFor(
            Service service,
            Func<Service, IEnumerable<ServiceRegistration>> registrationAccessor)
        {
            if (service is IServiceWithType st
                    && st.ServiceType.IsGenericType
                    && this._type == st.ServiceType.GetGenericTypeDefinition())
            {

                // get all non contravariant registration source 
                var originalRegistrations = registrationAccessor(service).ToArray();

                var components = _source
                                    // retrieve all contravariant registration source
                                    .RegistrationsFor(service, registrationAccessor)
                                    // Group will ensure having only a single registration of a activator limit type
                                    .GroupBy(this.GetTargetTypeDefinitionOrSelf)
                                    // exclude groups if autofac already resolved the same activator limit type
                                    .Where(o => !originalRegistrations.Select(oo => this.GetTargetTypeDefinitionOrSelf(oo.Registration)).Contains(o.Key))
                                    // taking the last is the default behavior for autofac, it can be improved
                                    .Select(o => o.Last());
                return components;
            }
            else
            {
                return Enumerable.Empty<IComponentRegistration>();
            }
        }

        private Type GetTargetTypeDefinitionOrSelf(IComponentRegistration componentRegistration)
        {
            return componentRegistration.Target.Activator.LimitType.IsGenericType ?
                        componentRegistration.Target.Activator.LimitType.GetGenericTypeDefinition()
                        : componentRegistration.Target.Activator.LimitType;
        }

        public bool IsAdapterForIndividualComponents => _source.IsAdapterForIndividualComponents;
    }

and I use it like this : 

            builder.RegisterSource(new ExplicitContravariantRegistrationSource(typeof(INotificationHandler<>)));


This issue is also discussed here : https://stackoverflow.com/questions/46464944/autofac-contravariance-and-resolving-open-generic-types/71010423#71010423

**Comment by lonix1** [comment]
Autofac fixed this issue in [6.2.0](https://github.com/autofac/Autofac/releases/tag/v6.2.0). But my generic `IRequestPostProcessor<TRequest, TResponse>` still doesn't work.

Did anyone else solve this problem with the latest version?
